src/hb-ot-layout-gsubgpos.hh

```diff
@@ -1175,8 +1192,14 @@ static inline bool match_backtrack (hb_ot_apply_context_t *c,
   skippy_iter.set_match_func (match_func, match_data, backtrack);
 
   for (unsigned int i = 0; i < count; i++)
-    if (!skippy_iter.prev ())
+  {
+    unsigned unsafe_from;
+    if (!skippy_iter.prev (&unsafe_from))
```

Comment:
IIUC here, and below in line 1226 is the logic for setting this flag, but I don't have enough knowledge of the codebase to tell what this does. What's the underlying logic of when this is set? What existing internal state of this `skippy_iter` is used to determine when to set the flag? Is there some sort of closure/pre-analysis done of potential ligature/other-shaping-sequence completions?

Reply:
No pre-analysis. Just every time it tries to match a ligature and fails, it marks up to the failure point as unsafe-to-concat.

Eg. if text is "hourglaxxyzdkjtlak" and the ligature we're trying to match is "hourglass_bottom", the matcher matches up the sequence "hourgla" and gives up upon seeing "x". It then marks unsafe-to-concat up to there, because up to the x, if something changes, there's a change of a ligature forming.

That's basically what it's doing.